mod/dfrn_confirm.php

```diff
@@ -136,270 +134,199 @@ function dfrn_confirm_post(App $a, $handsfree = null)
+		if (stristr($res, "<status") === false) {
+			// wrong xml! stop here!
+			Logger::log('Unexpected response posting to ' . $dfrn_confirm);
+			notice(L10n::t('Unexpected response from remote site: ') . EOL . htmlspecialchars($res) . EOL);
+			return;
+		}
 
-			$new_relation = $contact['rel'];
-			$writable = $contact['writable'];
+		$xml = XML::parseString($res);
+		$status = (int) $xml->status;
+		$message = XML::unescape($xml->message);   // human readable text of what may have gone wrong.
+		switch ($status) {
+			case 0:
+				info(L10n::t("Confirmation completed successfully.") . EOL);
+				break;
+			case 1:
+				// birthday paradox - generate new dfrn-id and fall through.
+				$new_dfrn_id = Strings::getRandomHex();
+				q("UPDATE contact SET `issued-id` = '%s' WHERE `id` = %d AND `uid` = %d",
+					DBA::escape($new_dfrn_id),
+					intval($contact_id),
+					intval($uid)
+				);
+
+			case 2:
+				notice(L10n::t("Temporary failure. Please wait and try again.") . EOL);
+				break;
+			case 3:
+				notice(L10n::t("Introduction failed or was revoked.") . EOL);
+				break;
+		}
 
-			if (in_array($protocol, [Protocol::DIASPORA, Protocol::ACTIVITYPUB])) {
-				if ($duplex) {
-					$new_relation = Contact::FRIEND;
-				} else {
-					$new_relation = Contact::FOLLOWER;
-				}
+		if (strlen($message)) {
+			notice(L10n::t('Remote site reported: ') . $message . EOL);
+		}
 
-				if ($new_relation != Contact::FOLLOWER) {
-					$writable = 1;
-				}
+		if (($status == 0) && $intro_id) {
+			$intro = DBA::selectFirst('intro', ['note'], ['id' => $intro_id]);
+			if (DBA::isResult($intro)) {
+				DBA::update('contact', ['reason' => $intro['note']], ['id' => $contact_id]);
 			}
 
+			// Success. Delete the notification.
 			DBA::delete('intro', ['id' => $intro_id]);
+		}
 
-			$fields = ['name-date' => DateTimeFormat::utcNow(),
-				'uri-date' => DateTimeFormat::utcNow(), 'addr' => $addr,
-				'notify' => $notify, 'poll' => $poll, 'blocked' => false,
-				'pending' => $pending, 'protocol' => $protocol,
-				'writable' => $writable, 'hidden' => $hidden, 'rel' => $new_relation];
-			DBA::update('contact', $fields, ['id' => $contact_id]);
+		if ($status != 0) {
+			return;
 		}
 
-		if (!DBA::isResult($r)) {
-			notice(L10n::t('Unable to set contact photo.') . EOL);
+		/*
+		 * We have now established a relationship with the other site.
+		 * Let's make our own personal copy of their profile photo so we don't have
+		 * to always load it from their site.
+		 *
+		 * We will also update the contact record with the nature and scope of the relationship.
+		 */
+		Contact::updateAvatar($contact['photo'], $uid, $contact_id);
+
+		Logger::log('dfrn_confirm: confirm - imported photos');
+
+		$new_relation = Contact::FOLLOWER;
+
+		if (($relation == Contact::SHARING) || ($duplex)) {
+			$new_relation = Contact::FRIEND;
 		}
 
+		if (($relation == Contact::SHARING) && ($duplex)) {
+			$duplex = 0;
+		}
+
+		$r = q("UPDATE `contact` SET `rel` = %d,
```

Comment:
`DBA::p()` instead of `q()`

Reply:
In that file I only had removed the non DFRN parts. I don't intent to completely rebuild it. And some nitpicking: it wouldn't be ```DBA::p``` but ```DBA::update``` :-)